crates/vm/levm/src/opcode_handlers/environment.rs

```diff
@@ -149,19 +152,37 @@ impl<'a> VM<'a> {
             return Ok(OpcodeResult::Continue { pc_increment: 1 });
         }
 
-        let mut data = vec![0u8; size];
+        // Happiest fast path, copy without an intermediate buffer because there is no need to pad 0s and also size doesn't overflow.
+        if let Some(calldata_offset_end) = calldata_offset.checked_add(size) {
+            if calldata_offset_end <= current_call_frame.calldata.len() {
+                #[expect(unsafe_code, reason = "bounds checked beforehand")]
+                let slice = unsafe {
+                    current_call_frame
+                        .calldata
+                        .get_unchecked(calldata_offset..calldata_offset_end)
+                };
+                current_call_frame.memory.store_data(dest_offset, slice)?;
+
+                return Ok(OpcodeResult::Continue { pc_increment: 1 });
+            }
+        }
 
-        if calldata_offset <= current_call_frame.calldata.len() {
-            for (i, byte) in current_call_frame
+        let mut data = vec![0u8; size];
```

Comment:
Isn't size bounded to 32? Maybe use an array instead to avoid the allocation.

Reply:
Ah no, this one is variable. I thought it was the same function as above.

Reply:
That's only the case for `CALLDATALOAD`. This is for `CALLDATACOPY` which takes an arbitrary size.